File: Storyboard/frontend/api/storyboard.ts
/**
 * 故事板功能 API 封装
 * - 分镜头脚本生成：GLM-5
 * - 故事板图片生成：nano-banana2
 * - 角色设计图生成：nano-banana2
 */

import { createNanoBanana2API } from './nanobanana2'
import type {
  StoryboardScript,
  StoryboardStyle,
  StoryboardCharacter,
  DialogueLine,
  parseAndNormalizeScript,
} from '@/stores/storyboardStore'

// GLM API 配置
const GLM_API_URL = 'https://open.bigmodel.cn/api/paas/v4/chat/completions'
const GLM_MODEL = 'glm-5'  // GLM-5 旗舰模型

// 风格对应的提示词后缀
const STYLE_PROMPTS: Record<StoryboardStyle, string> = {
  comic: 'comic book style, black and white manga panel, ink drawing, screentone, clear outlines, storyboard frame',
  realistic: 'cinematic storyboard sketch, realistic film style, professional storyboard artist, pencil sketch',
  anime: 'anime storyboard style, Japanese animation, clean lines, vibrant colors, anime production art',
  watercolor: 'watercolor storyboard style, soft brush strokes, artistic illustration, hand-painted look',
}

// 景别对应的英文提示词
const SHOT_TYPE_PROMPTS: Record<string, string> = {
  远景: 'wide establishing shot',
  全景: 'full shot',
  中景: 'medium shot',
  近景: 'close-up shot',
  特写: 'extreme close-up',
  大特写: 'extreme close-up detail',
}

/**
 * 生成分镜头脚本（增强版）
 */
export async function generateStoryboardScript(
  apiKey: string,
  inputText: string,
  onProgress?: (progress: number) => void
): Promise<StoryboardScript> {
  onProgress?.(10)

  const systemPrompt = `你是一个专业的分镜头脚本编剧。根据用户提供的文案内容，生成结构化的分镜头脚本。

## 输出要求

### 1. 剧本分析
- 分析文案，拆分为 4-9 个镜头
- 总时长控制在 3-4 分钟
- 提取剧本梗概（synopsis）

### 2. 角色提取（仅主要角色，≤5人）
为每个主要角色提供详细描述：
- 基本信息：姓名、年龄、性别、角色定位（protagonist主角/supporting配角/minor龙套）
- 外观特征：身高、体型、发型发色、瞳色、肤色、显著特征
- 服装：主要服装、配饰、主色调
- 性格：性格特点、习惯动作、说话风格

### 3. 场景设计
每个镜头包含：
- 时长、景别、画面描述、镜头运动
- 结构化对白列表（角色ID、角色名、台词、情绪、语气、语速、停顿、舞台指示）
- 旁白（如有）

### 4. 对白情绪标注（为TTS合成准备）
- 情绪类型：neutral, happy, sad, angry, surprised, fearful, disgusted, contemptuous, excited, calm
- 语气类型：normal, whisper, shout, sarcastic, gentle, stern, playful, hesitant, confident
- 情绪强度：1-10（整数）
- 语速：0.5-2.0（1.0为正常，保留一位小数）
- 停顿：秒数（0.5的倍数）

必须严格按以下 JSON 格式输出，不要有任何其他内容：
{
  "title": "分镜头标题",
  "totalDuration": "3:30",
  "synopsis": "剧本梗概，简述故事内容和主题",
  "characters": [
    {
      "id": "char_1",
      "name": "角色名",
      "role": "protagonist",
      "description": "简短的角色描述",
      "appearance": {
        "age": "25岁",
        "gender": "男",
        "height": "180cm",
        "build": "健壮",
        "hairColor": "黑色",
        "hairStyle": "短发",
        "eyeColor": "深棕色",
        "skinTone": "小麦色",
        "distinctiveFeatures": ["剑眉", "左眉有疤痕"]
      },
      "costume": {
        "mainOutfit": "青色长袍",
        "accessories": ["玉佩", "草鞋"],
        "colors": ["青色", "白色"]
      },
      "personality": {
        "traits": ["沉稳", "内敛", "坚韧"],
        "mannerisms": ["思考时习惯摸下巴"],
        "speakingStyle": "语速平稳，声音低沉"
      }
    }
  ],
  "scenes": [
    {
      "id": 1,
      "duration": "0:30",
      "shotType": "中景",
      "description": "详细的画面描述，用于AI绘图",
      "camera": "固定镜头",
      "dialogues": [
        {
          "characterId": "char_1",
          "characterName": "陈平安",
          "text": "对白内容",
          "emotion": "calm",
          "emotionIntensity": 7,
          "tone": "gentle",
          "speed": 1.0,
          "pause": 0.5,
          "stageDirection": "缓缓转身"
        }
      ],
      "narrator": "旁白内容（可选）"
    }
  ],
  "allDialogues": [
    {
      "characterId": "char_1",
      "characterName": "陈平安",
      "text": "所有对白汇总",
      "emotion": "calm",
      "emotionIntensity": 7,
      "tone": "gentle",
      "speed": 1.0,
      "pause": 0.5,
      "stageDirection": ""
    }
  ]
}

景别选项：远景、全景、中景、近景、特写、大特写
镜头运动：固定镜头、推、拉、摇、移、跟、升降

注意：
1. characters 数组只包含有对白的主要角色，最多5个
2. dialogues 数组中的 characterId 必须对应 characters 中的 id
3. allDialogues 是所有场景中对白的汇总，便于后期TTS处理
4. 确保输出的 JSON 格式正确，可以被 JSON.parse() 解析`

  const response = await fetch(GLM_API_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model: GLM_MODEL,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: `请根据以下文案生成分镜头脚本：\n\n${inputText}` },
      ],
      temperature: 0.7,
      max_tokens: 8192,  // 增加输出长度以容纳详细数据
    }),
  })

  onProgress?.(50)

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new Error(errorData.error?.message || `API 请求失败: ${response.status}`)
  }

  const data = await response.json()
  const content = data.choices?.[0]?.message?.content || ''

  onProgress?.(80)

  // 解析 JSON
  try {
    // 提取 JSON 内容（支持 markdown 代码块）
    let jsonStr = content

    // 尝试提取 ```json ... ``` 中的内容
    const jsonBlockMatch = content.match(/```(?:json)?\s*([\s\S]*?)```/)
    if (jsonBlockMatch) {
      jsonStr = jsonBlockMatch[1].trim()
    } else {
      // 尝试提取纯 JSON 对象
      const jsonMatch = content.match(/\{[\s\S]*\}/)
      if (!jsonMatch) {
        throw new Error('未能解析出 JSON 格式的脚本')
      }
      jsonStr = jsonMatch[0]
    }

    const parsed = JSON.parse(jsonStr)

    // 使用规范化函数填充默认值
    const script = normalizeScriptData(parsed)

    onProgress?.(100)
    return script
  } catch (parseError) {
    console.error('[Storyboard] Parse JSON error:', parseError)
    console.error('[Storyboard] Raw content:', content.substring(0, 500))
    throw new Error('分镜头脚本解析失败，请重试')
  }
}

/**
 * 规范化脚本数据，为缺失字段提供默认值
 */
function normalizeScriptData(data: any): StoryboardScript {
  const characters: StoryboardCharacter[] = (data.characters || []).map((c: any, i: number) => ({
    id: c.id || `char_${i + 1}`,
    name: c.name || `角色${i + 1}`,
    role: c.role || 'supporting',
    description: c.description || '',
    appearance: {
      age: c.appearance?.age || '未知',
      gender: c.appearance?.gender || '未知',
      height: c.appearance?.height || '中等',
      build: c.appearance?.build || '普通',
      hairColor: c.appearance?.hairColor || '黑色',
      hairStyle: c.appearance?.hairStyle || '普通',
      eyeColor: c.appearance?.eyeColor || '黑色',
      skinTone: c.appearance?.skinTone || '正常',
      distinctiveFeatures: c.appearance?.distinctiveFeatures || [],
    },
    costume: {
      mainOutfit: c.costume?.mainOutfit || '普通服装',
      accessories: c.costume?.accessories || [],
      colors: c.costume?.colors || [],
    },
    personality: {
      traits: c.personality?.traits || [],
      mannerisms: c.personality?.mannerisms || [],
      speakingStyle: c.personality?.speakingStyle || '正常',
    },
    referenceImageUrl: c.referenceImageUrl,
  }))

  const scenes = (data.scenes || []).map((s: any) => ({
    id: s.id ?? 0,
    duration: s.duration || '0:30',
    shotType: s.shotType || '中景',
    description: s.description || '',
    camera: s.camera || '固定镜头',
    dialogues: (s.dialogues || []).map((d: any) => ({
      characterId: d.characterId || '',
      characterName: d.characterName || '未知',
      text: d.text || '',
      emotion: d.emotion || 'neutral',
      emotionIntensity: d.emotionIntensity ?? 5,
      tone: d.tone || 'normal',
      speed: d.speed ?? 1.0,
      pause: d.pause ?? 0,
      stageDirection: d.stageDirection || '',
    })),
    narrator: s.narrator || '',
  }))

  const allDialogues: DialogueLine[] = (data.allDialogues || scenes.flatMap((s: any) => s.dialogues)).map((d: any) => ({
    characterId: d.characterId || '',
    characterName: d.characterName || '未知',
    text: d.text || '',
    emotion: d.emotion || 'neutral',
    emotionIntensity: d.emotionIntensity ?? 5,
    tone: d.tone || 'normal',
    speed: d.speed ?? 1.0,
    pause: d.pause ?? 0,
    stageDirection: d.stageDirection || '',
  }))

  return {
    title: data.title || '未命名分镜头',
    totalDuration: data.totalDuration || '3:00',
    synopsis: data.synopsis || '',
    characters,
    scenes,
    allDialogues,
  }
}

/**
 * 构建故事板图片提示词（增强版）
 */
function buildStoryboardPrompt(
  scene: { description: string; shotType: string },
  style: StoryboardStyle,
  characterDescriptions?: string[]
): string {
  const stylePrompt = STYLE_PROMPTS[style]
  const shotPrompt = SHOT_TYPE_PROMPTS[scene.shotType] || 'medium shot'

  let prompt = `${scene.description}, ${shotPrompt}, ${stylePrompt}, high quality, detailed`

  if (characterDescriptions && characterDescriptions.length > 0) {
    prompt += `, characters: ${characterDescriptions.join(', ')}`
  }

  return prompt
}

/**
 * 批量生成故事板图片（并行）
 */
export async function generateStoryboardImages(
  apiKey: string,
  script: StoryboardScript,
  style: StoryboardStyle,
  onProgress?: (current: number, total: number, imageUrl?: string) => void
): Promise<string[]> {
  const scenes = script.scenes
  const total = scenes.length
  const images: string[] = new Array(total).fill('')
  let completed = 0

  // 构建角色描述用于图片生成
  const characterDescs = script.characters.map((c) =>
    `${c.name}: ${c.appearance.age}, ${c.appearance.gender}, ${c.appearance.build}, ${c.appearance.hairColor} ${c.appearance.hairStyle}, ${c.costume.mainOutfit}`
  )

  // 并行生成所有图片
  const tasks = scenes.map(async (scene, i) => {
    const api = createNanoBanana2API(apiKey)
    const prompt = buildStoryboardPrompt(scene, style, characterDescs)

    try {
      const urls = await api.generateImageWithProgress({
        prompt,
        size: '1K',
        aspectRatio: '16:9',
      }, () => {})

      if (urls.length > 0) {
        images[i] = urls[0]
        completed++
        onProgress?.(completed, total, urls[0])
      }
    } catch (error) {
      console.error(`[Storyboard] Generate image ${i + 1} failed:`, error)
      completed++
      onProgress?.(completed, total)
    }
  })

  await Promise.all(tasks)

  // 过滤空结果
  return images.filter(Boolean)
}

/**
 * 构建角色设计图提示词（每个角色一张组合图）
 */
function buildCharacterDesignPrompt(character: StoryboardCharacter): string {
  const { appearance, costume, name, personality } = character

  return `character design sheet, ${name},
${appearance.age}, ${appearance.gender}, ${appearance.height}, ${appearance.build},
${appearance.hairColor} ${appearance.hairStyle}, ${appearance.eyeColor} eyes, ${appearance.skinTone} skin,
${appearance.distinctiveFeatures.join(', ') || 'no distinctive features'},
wearing ${costume.mainOutfit}, ${costume.accessories.join(', ') || 'no accessories'},
color scheme: ${costume.colors.join(', ') || 'default'},

layout: 3 full body standing poses (front view, side view, back view) on the left,
4 face close-up expressions (neutral, happy, serious, surprised) on the right,

personality hints: ${personality.traits.join(', ')},
${personality.mannerisms.join(', ')},

white background, professional character design, clean lines, detailed, high quality,
reference sheet style, multiple views, character turnaround`
}

/**
 * 批量生成角色设计图
 * 每个角色生成一张组合图（3站姿 + 4表情）
 */
export async function generateCharacterDesigns(
  apiKey: string,
  characters: StoryboardCharacter[],
  onProgress?: (current: number, total: number, imageUrl?: string, characterId?: string) => void
): Promise<Array<{ characterId: string; characterName: string; imageUrl: string }>> {
  // 只为主要角色生成设计图
  const mainCharacters = characters.filter(c =>
    c.role === 'protagonist' || c.role === 'supporting'
  ).slice(0, 5)  // 最多5个角色

  if (mainCharacters.length === 0) {
    return []
  }

  const total = mainCharacters.length
  const results: Array<{ characterId: string; characterName: string; imageUrl: string }> = []
  let completed = 0

  // 并行生成所有角色设计图
  const tasks = mainCharacters.map(async (character) => {
    const api = createNanoBanana2API(apiKey)
    const prompt = buildCharacterDesignPrompt(character)

    try {
      const urls = await api.generateImageWithProgress({
        prompt,
        size: '1K',
        aspectRatio: '16:9',  // 组合图使用宽画幅
      }, () => {})

      if (urls.length > 0) {
        results.push({
          characterId: character.id,
          characterName: character.name,
          imageUrl: urls[0],
        })
        completed++
        onProgress?.(completed, total, urls[0], character.id)
      } else {
        completed++
        onProgress?.(completed, total)
      }
    } catch (error) {
      console.error(`[Storyboard] Generate character ${character.name} failed:`, error)
      completed++
      onProgress?.(completed, total)
    }
  })

  await Promise.all(tasks)

  return results
}

/**
 * 一键生成所有内容
 * @param textApiKey 智谱 API Key，用于生成分镜头脚本
 * @param imageApiKey 速创 API Key，用于生成图片
 */
export async function generateAll(
  textApiKey: string,
  imageApiKey: string,
  inputText: string,
  style: StoryboardStyle,
  onTaskChange?: (task: 'script' | 'storyboard' | 'character') => void,
  onProgress?: (progress: number, message?: string) => void
): Promise<{
  script: StoryboardScript
  storyboardImages: string[]
  characterDesigns: Array<{ characterId: string; characterName: string; imageUrl: string }>
}> {
  // 任务1：生成分镜头脚本（使用智谱 API）
  onTaskChange?.('script')
  onProgress?.(0, '正在分析文案生成分镜头脚本...')
  const script = await generateStoryboardScript(textApiKey, inputText, (p) => {
    onProgress?.(p * 0.3, `正在生成分镜头脚本...${p}%`)
  })

  // 任务2：生成故事板图片（使用速创 API）
  onTaskChange?.('storyboard')
  onProgress?.(30, `正在生成${script.scenes.length}张故事板图片...`)
  const storyboardImages = await generateStoryboardImages(imageApiKey, script, style, (current, total, url) => {
    const progress = 30 + (current / total) * 40
    onProgress?.(progress, `正在生成故事板图片 ${current}/${total}`)
  })

  // 任务3：生成角色设计图（使用速创 API）
  onTaskChange?.('character')
  const characterCount = script.characters.filter(c => c.role === 'protagonist' || c.role === 'supporting').length
  onProgress?.(70, `正在生成${characterCount}个角色设计图...`)
  const characterDesigns = await generateCharacterDesigns(
    imageApiKey,
    script.characters,
    (current, total, url, characterId) => {
      const progress = 70 + (current / total) * 30
      onProgress?.(progress, `正在生成角色设计图 ${current}/${total}`)
    }
  )

  onProgress?.(100, '全部生成完成！')

  return {
    script,
    storyboardImages,
    characterDesigns,
  }
}

File: Storyboard/backend/schemas/storyboard.py
"""
Nano2 API - AI Image Generation Service

Copyright ©2026 IoTchange (外星动物/常智)
All Rights Reserved.

Author: 外星动物（常智）IoTchange
Email: 14455975@qq.com
Version: 2.2.2

本软件著作权归作者 IoTchange 完整所有。
商用需授权，开源使用需标明作者。
"""

from enum import Enum
from typing import Optional, Any

from pydantic import BaseModel, Field


class StoryboardStyle(str, Enum):
    """故事板风格"""
    
    CINEMATIC = "cinematic"      # 电影风格
    ANIME = "anime"              # 动漫风格
    COMIC = "comic"              # 漫画风格
    REALISTIC = "realistic"      # 写实风格
    WATERCOLOR = "watercolor"    # 水彩风格
    SKETCH = "sketch"            # 素描风格


class AspectRatio(str, Enum):
    """画面比例"""
    
    RATIO_16_9 = "16:9"          # 宽屏
    RATIO_4_3 = "4:3"            # 标准
    RATIO_1_1 = "1:1"            # 正方形
    RATIO_9_16 = "9:16"          # 竖屏
    RATIO_21_9 = "21:9"          # 超宽


class SceneInfo(BaseModel):
    """场景信息"""
    
    index: int = Field(..., ge=0, description="场景索引")
    shot_type: str = Field(..., description="镜头类型")
    description: str = Field(..., description="场景描述")
    dialogue: Optional[str] = Field(None, description="对白")
    action: Optional[str] = Field(None, description="动作描述")
    camera_movement: Optional[str] = Field(None, description="镜头运动")
    duration: Optional[float] = Field(None, ge=0, description="时长（秒）")
    notes: Optional[str] = Field(None, description="备注")
    prompt: str = Field(..., description="图像生成提示词")
    image_url: Optional[str] = Field(None, description="生成的图像URL")
    
    class Config:
        json_schema_extra = {
            "example": {
                "index": 0,
                "shot_type": "中景",
                "description": "一个年轻人站在城市天台上，望着远方的夕阳",
                "dialogue": "我们终究会到达那里",
                "action": "缓缓转身",
                "camera_movement": "缓慢推进",
                "duration": 5.0,
                "notes": "注意光线要柔和",
                "prompt": "一个年轻人站在城市天台上，望着远方的夕阳，电影感，金色光芒",
                "image_url": None
            }
        }


class StoryboardScriptRequest(BaseModel):
    """故事板脚本生成请求"""
    
    story: str = Field(..., min_length=10, max_length=5000, description="故事文本")
    style: StoryboardStyle = Field(default=StoryboardStyle.CINEMATIC, description="故事板风格")
    aspect_ratio: AspectRatio = Field(default=AspectRatio.RATIO_16_9, description="画面比例")
    num_scenes: int = Field(default=6, ge=3, le=20, description="场景数量")
    include_dialogue: bool = Field(default=True, description="是否包含对白")
    include_camera: bool = Field(default=True, description="是否包含镜头信息")
    language: str = Field(default="zh", description="输出语言")
    
    class Config:
        json_schema_extra = {
            "example": {
                "story": "一个年轻的冒险家踏上了寻找传说中失落城市的旅程。在穿越丛林、沙漠和高山后，他终于找到了那座被遗忘的城市，发现了比黄金更珍贵的东西——真正的友谊。",
                "style": "cinematic",
                "aspect_ratio": "16:9",
                "num_scenes": 6,
                "include_dialogue": True,
                "include_camera": True,
                "language": "zh"
            }
        }


class StoryboardScriptResponse(BaseModel):
    """故事板脚本生成响应"""

    task_id: int = Field(..., description="任务ID")
    status: str = Field(..., description="任务状态")
    title: Optional[str] = Field(None, description="故事标题")
    synopsis: Optional[str] = Field(None, description="故事摘要")
    style: StoryboardStyle = Field(..., description="故事板风格")
    aspect_ratio: AspectRatio = Field(..., description="画面比例")
    characters: list[dict] = Field(default_factory=list, description="角色列表")
    scenes: list[SceneInfo] = Field(default_factory=list, description="场景列表")
    total_duration: Optional[float] = Field(None, description="总时长（秒）")
    created_at: str = Field(..., description="创建时间")
    
    class Config:
        json_schema_extra = {
            "example": {
                "task_id": 1,
                "status": "completed",
                "title": "寻找失落的城市",
                "synopsis": "一个关于冒险与友谊的故事",
                "style": "cinematic",
                "aspect_ratio": "16:9",
                "scenes": [
                    {
                        "index": 0,
                        "shot_type": "远景",
                        "description": "日出时分，一座繁华的城市全景",
                        "prompt": "日出时分的城市全景，金色阳光洒落，电影感",
                        "image_url": None
                    }
                ],
                "total_duration": 30.0,
                "created_at": "2026-01-01T00:00:00"
            }
        }


class StoryboardImagesRequest(BaseModel):
    """故事板图片生成请求"""

    storyboard_id: int = Field(..., description="故事板ID")
    scene_indices: Optional[list[int]] = Field(None, description="指定场景索引（不填则生成全部）")
    image_size: str = Field(default="1024x1024", description="图像尺寸")
    enhance_prompts: bool = Field(default=True, description="是否增强提示词")
    parallel: bool = Field(default=False, description="是否并行生成")
    auto_save: bool = Field(default=False, description="是否自动保存到资产库")

    class Config:
        json_schema_extra = {
            "example": {
                "storyboard_id": 1,
                "scene_indices": [0, 1, 2],
                "image_size": "1024x1024",
                "enhance_prompts": True,
                "parallel": False,
                "auto_save": False
            }
        }


class SceneImageStatus(BaseModel):
    """场景图片状态"""
    
    scene_index: int = Field(..., description="场景索引")
    status: str = Field(..., description="状态")
    task_id: Optional[int] = Field(None, description="任务ID")
    image_url: Optional[str] = Field(None, description="图像URL")
    error: Optional[str] = Field(None, description="错误信息")
    
    class Config:
        json_schema_extra = {
            "example": {
                "scene_index": 0,
                "status": "completed",
                "task_id": 1,
                "image_url": "https://example.com/images/scene_0.png",
                "error": None
            }
        }


class StoryboardImagesResponse(BaseModel):
    """故事板图片生成响应"""

    storyboard_id: int = Field(..., description="故事板ID")
    total_scenes: int = Field(..., description="总场景数")
    completed: int = Field(..., description="已完成数")
    scenes: list[SceneImageStatus] = Field(..., description="场景状态列表")
    saved_to_library: bool = Field(default=False, description="是否已保存到资产库")

    class Config:
        json_schema_extra = {
            "example": {
                "storyboard_id": 1,
                "total_scenes": 6,
                "completed": 6,
                "scenes": [
                    {
                        "scene_index": 0,
                        "status": "completed",
                        "task_id": None,
                        "image_url": "https://picsum.photos/seed/1234/1024/1024",
                        "error": None
                    }
                ],
                "saved_to_library": False
            }
        }


class StoryboardDetail(BaseModel):
    """故事板详情"""

    id: int = Field(..., description="故事板ID")
    title: str = Field(..., description="标题")
    synopsis: Optional[str] = Field(None, description="摘要")
    style: StoryboardStyle = Field(..., description="风格")
    aspect_ratio: AspectRatio = Field(..., description="画面比例")
    scenes: list[SceneInfo] = Field(..., description="场景列表")
    characters: list["CharacterDesignInfo"] = Field(default_factory=list, description="角色设计列表")
    total_duration: Optional[float] = Field(None, description="总时长")
    created_at: str = Field(..., description="创建时间")
    updated_at: str = Field(..., description="更新时间")

    class Config:
        json_schema_extra = {
            "example": {
                "id": 1,
                "title": "寻找失落的城市",
                "synopsis": "一个关于冒险与友谊的故事",
                "style": "cinematic",
                "aspect_ratio": "16:9",
                "scenes": [],
                "characters": [],
                "total_duration": 30.0,
                "created_at": "2026-01-01T00:00:00",
                "updated_at": "2026-01-01T00:00:00"
            }
        }


# ============ 角色设计相关 Schema ============

class CharacterAppearance(BaseModel):
    """角色外貌"""

    age: str = Field(..., description="年龄段")
    gender: str = Field(..., description="性别")
    build: str = Field(..., description="体型")
    height: str = Field(..., description="身高")
    hair_color: str = Field(..., description="发色")
    hair_style: str = Field(..., description="发型")
    eye_color: str = Field(..., description="瞳色")
    skin_tone: str = Field(..., description="肤色")
    distinctive_features: list[str] = Field(default_factory=list, description="特征")


class CharacterCostume(BaseModel):
    """角色服装"""

    main_outfit: str = Field(..., description="主要服装")
    accessories: list[str] = Field(default_factory=list, description="配饰")
    colors: list[str] = Field(default_factory=list, description="颜色")


class CharacterPersonality(BaseModel):
    """角色性格"""

    traits: list[str] = Field(default_factory=list, description="性格特点")
    mannerisms: list[str] = Field(default_factory=list, description="习惯动作")


class CharacterDesignInfo(BaseModel):
    """角色设计信息"""

    id: str = Field(..., description="角色ID")
    name: str = Field(..., description="角色名")
    role: str = Field(..., description="角色类型：protagonist/supporting/minor")
    description: str = Field(..., description="角色描述")
    appearance: CharacterAppearance = Field(..., description="外貌")
    costume: CharacterCostume = Field(..., description="服装")
    personality: CharacterPersonality = Field(default_factory=CharacterPersonality, description="性格")
    design_prompt: Optional[str] = Field(None, description="设计图提示词")
    image_url: Optional[str] = Field(None, description="设计图URL")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "char_1",
                "name": "小明",
                "role": "protagonist",
                "description": "一个充满冒险精神的年轻人",
                "appearance": {
                    "age": "20岁左右",
                    "gender": "男",
                    "build": "身材匀称",
                    "height": "175cm",
                    "hair_color": "黑色",
                    "hair_style": "短发",
                    "eye_color": "棕色",
                    "skin_tone": "健康小麦色",
                    "distinctive_features": ["左眼角有颗小痣"]
                },
                "costume": {
                    "main_outfit": "蓝色探险夹克，卡其色工装裤",
                    "accessories": ["背包", "指南针"],
                    "colors": ["蓝色", "卡其色"]
                },
                "personality": {
                    "traits": ["勇敢", "乐观", "善良"],
                    "mannerisms": ["思考时喜欢摸下巴"]
                },
                "design_prompt": None,
                "image_url": None
            }
        }


class CharacterExtractRequest(BaseModel):
    """角色提取请求"""

    story: str = Field(..., min_length=10, max_length=10000, description="故事文本")
    style: StoryboardStyle = Field(default=StoryboardStyle.CINEMATIC, description="故事板风格")
    max_characters: int = Field(default=5, ge=1, le=10, description="最大角色数")


class CharacterExtractResponse(BaseModel):
    """角色提取响应"""

    characters: list[CharacterDesignInfo] = Field(..., description="角色列表")


class CharacterDesignRequest(BaseModel):
    """角色设计图生成请求"""

    characters: list[CharacterDesignInfo] = Field(..., description="角色列表")
    style: StoryboardStyle = Field(default=StoryboardStyle.CINEMATIC, description="故事板风格")


class CharacterDesignResponse(BaseModel):
    """角色设计图生成响应"""

    characters: list[CharacterDesignInfo] = Field(..., description="角色设计结果")
    completed: int = Field(..., description="已完成数")
    total: int = Field(..., description="总数")
